src/scenic/core/regions.py

```diff
@@ -281,883 +306,2816 @@ def __hash__(self):
+        y, x = numpy.where(self.grid == 0)
+        points = [self.gridToPoint(point) for point in zip(x, y)]
+        super().__init__(name, points, orientation=orientation)
+
+    def gridToPoint(self, gp):
+        x, y = gp
+        return ((self.Ax * x) + self.Bx, (self.Ay * y) + self.By)
+
+    def pointToGrid(self, point):
+        x, y, z = point
+        x = (x - self.Bx) / self.Ax
+        y = (y - self.By) / self.Ay
+        nx = int(round(x))
+        if nx < 0 or nx >= self.sizeX:
+            return None
+        ny = int(round(y))
+        if ny < 0 or ny >= self.sizeY:
+            return None
+        return (nx, ny)
+
+    def containsPoint(self, point):
+        gp = self.pointToGrid(point)
+        if gp is None:
+            return False
+        x, y = gp
+        return (self.grid[y, x] == 0)
+
+    def containsObject(self, obj):
+        # TODO improve this procedure!
+        # Fast check
+        for c in obj.corners:
+            if not self.containsPoint(c):
+                return False
+        # Slow check
+        gps = [self.pointToGrid(corner) for corner in obj.corners]
+        x, y = zip(*gps)
+        minx, maxx = findMinMax(x)
+        miny, maxy = findMinMax(y)
+        for x in range(minx, maxx+1):
+            for y in range(miny, maxy+1):
+                p = self.gridToPoint((x, y))
+                if self.grid[y, x] == 1 and obj.containsPoint(p):
+                    return False
+        return True
+
+    def containsRegionInner(self, reg, tolerance):
+        raise NotImplementedError
+
+###################################################################################################
+# View Regions
+###################################################################################################
+
+class DefaultViewRegion(MeshVolumeRegion):
+    """ The default view region shape.
+
+    The default view region can take several forms, depending on the viewAngles parameter:
+
+    * Case 1:       viewAngles[0] = 360 degrees
+
+      * Case 1.a:   viewAngles[1] = 180 degrees         => Full Sphere View Region
+      * Case 1.b:   viewAngles[1] < 180 degrees         => Sphere - (Cone + Cone) (Cones on z axis expanding from origin)
+
+    * Case 2:       viewAngles[0] = 180 degrees
+
+      * Case 2.a:   viewAngles[1] = 180 degrees         => Hemisphere View Region
+      * Case 2.b:   viewAngles[1] < 180 degrees         => Hemisphere - (Cone + Cone) (Cones on appropriate hemispheres)
+
+    * Case 3:       viewAngles[1] = 180 degrees
+
+      * Case 3.a:   viewAngles[0] < 180 degrees         => Sphere intersected with Pyramid View Region
+      * Case 3.b:   viewAngles[0] > 180 degrees         => Sphere - Backwards Pyramid View Region
+
+    * Case 4:       viewAngles[0 & 1] < 180             => Capped Pyramid View Region
+    * Case 5:       viewAngles[0] > 180, Altitude < 180 => (Sphere - (Cone + Cone) (Cones on appropriate hemispheres)) - Backwards Capped Pyramid View Region
```

Comment:
It would be easier to see that these cases are exhaustive if you used the order 1, 5, 3, 2, 4.